### PR TITLE
Add Html Log listener

### PR DESCRIPTION
**Pull request by fenxu**
1. Add html log listner
2. Adjust the log style for markdown rewriter

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            var dir = Path.GetDirectoryName(reportPath);
+            if (!string.IsNullOrEmpty(dir))
+            {
+                Directory.CreateDirectory(dir);
+            }
+            _writer = new StreamWriter(reportPath, true);
+            _writer.WriteLine("<html>");
+            _writer.WriteLine(@"<body lang=""en-us"">");
+            _writer.WriteLine(@"<h1 id=""azure_migration_report"">Azure Content Pre Migration Report</h1>");
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs]: why `azure`?

Comment (fenxu) [inline src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs]: ...Remove that for common

---
In reply to: [74730597](https://github.com/dotnet/docfx/pull/605#discussion_r74730597) [](ancestors = 74730597)

File: src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        }
+#endif
+
+        public HtmlLogListener(StreamWriter writer)
+        {
+            if (writer == null)
+            {
+                throw new ArgumentNullException(nameof(writer));
+            }
+            _writer = writer;
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs]: no `<html>...`, dispose seems always append `...</html>`

Comment (fenxu) [inline src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs]: Will add here

---
In reply to: [74730872](https://github.com/dotnet/docfx/pull/605#discussion_r74730872) [](ancestors = 74730872)

File: src/Microsoft.DocAsCode.AzureMarkdownRewriters/AzureMigrationEngineBuilder.cs [hunk 3]
```diff
@@ -275,7 +278,35 @@ private string GenerateAzureLinkHref(IMarkdownContext context, string href, stri
...
+
+            var currentFilePath = (string)context.Variables["path"];
+            var currentFolderPath = Path.GetDirectoryName(currentFilePath);
+
+            try
+            {
+                var nonMdExpectedPath = Path.Combine(currentFolderPath, nonMdHref);
+                if (!File.Exists(nonMdExpectedPath))
+                {
+                    Logger.LogWarning($"Can't find resorece reference: {nonMdHref}. Raw: {rawMarkdown}.", null, currentFilePath, line);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.AzureMarkdownRewriters/AzureMigrationEngineBuilder.cs]: >resorece [](start = 51, length = 8)

resource?

File: src/Microsoft.DocAsCode.AzureMarkdownRewriters/AzureMigrationEngineBuilder.cs [hunk 4]
```diff
@@ -275,7 +278,35 @@ private string GenerateAzureLinkHref(IMarkdownContext context, string href, stri
...
+                }
+                return nonMdHref;
+            }
+            catch (NotSupportedException nse)
+            {
+                Logger.LogError($"FixNonMdRelativeFileHref can't be apply on reference: {nonMdHref}. Raw: {rawMarkdown}. Exception: {nse.Message}", null, currentFilePath, line);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.AzureMarkdownRewriters/AzureMigrationEngineBuilder.cs]: >apply [](start = 68, length = 5)

applied

File: src/Microsoft.DocAsCode.AzureMarkdownRewriters/AzureMigrationEngineBuilder.cs [hunk 5]
```diff
@@ -275,7 +278,35 @@ private string GenerateAzureLinkHref(IMarkdownContext context, string href, stri
...
+            try
+            {
+                var nonMdExpectedPath = Path.Combine(currentFolderPath, nonMdHref);
+                if (!File.Exists(nonMdExpectedPath))
+                {
+                    Logger.LogWarning($"Can't find resorece reference: {nonMdHref}. Raw: {rawMarkdown}.", null, currentFilePath, line);
+                }
+                return nonMdHref;
+            }
+            catch (NotSupportedException nse)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.AzureMarkdownRewriters/AzureMigrationEngineBuilder.cs]: >NotSupportedException [](start = 19, length = 21)

Seems no function in the try scope will throw this type of exception?

File: src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            WriteCommonHeader();
+        }
+
+        public void WriteCommonHeader()
+        {
+            _writer.WriteLine("<html>");
+            _writer.WriteLine(@"<body lang=""en-us"">");
+            _writer.WriteLine(@"<h1 id=""Report"">Report</h1>");
+            _writer.WriteLine(@"<table border=""1"">");
+            _writer.WriteLine(@"<tr><td>Severity</td><td>Message</td><td>File</td><td>Line</td><td>Time</td></tr>");
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs]: `td` -> `th`?

File: src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            {
+                Severity = GetSeverity(level),
+                Message = message,
+                Source = phase,
+                File = file,
+                Line = line,
+                DateTime = DateTime.UtcNow
+            };
+
+            _writer.WriteLine($@"<tr><td>{reportItem.Severity}</td><td>{Escape(reportItem.Message)}</td><td>{reportItem.File}</td><td>{reportItem.Line}</td><td>{reportItem.DateTime}</td></tr>");
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs]: also escape `File` and `Line`

**Comment by vwxyzh** [comment]
:+1:

File: src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs (hunk 7 above)
Comment (fenxu) [inline src/Microsoft.DocAsCode.Common/Loggers/HtmlLogListener.cs]: Escape File. Line is not necessary to escape.

---
In reply to: [74915055](https://github.com/dotnet/docfx/pull/605#discussion_r74915055) [](ancestors = 74915055)

**Pull request merged**